dlib/cuda/tensor.h

```diff
@@ -50,6 +50,15 @@ namespace dlib
         long long nr() const { return m_nr; }
         long long nc() const { return m_nc; }
         size_t size() const { return m_size; }
+        size_t index(
+            const long sample,
+            const long k,
+            const long r,
+            const long c
+        ) const
+        {
+            return ((sample * m_k + k) * m_nr + r) * m_nc + c;
+        }
```

Comment:
This really doesn't belong here.  If we wanted to do this it would be better to have tensor work like ```
cpu_tensor_view view = my_tensor.cpu_view();
view(sample,k,r,c) = 123;
```
The point of `cpu_tensor_view` is to incur the cost of calling `host()` once and to provide nicer indexed access.  Could also have a `gpu_tensor_view` that's the same thing basically but calls `device()` when created.  Then that could be passed to cuda kernels instead of pointers and ints.

Reply:
I see, maybe for another PR, then... for now I will just make a static function `tensor_index` in the `misc.h` file and use it across loss functions, instead of each one having their own. Would that be OK?

Reply:
Yeah, that's fine.